src/pnm/decoder.rs

```diff
@@ -18,6 +18,171 @@ use crate::utils;
+    InvalidDepth {
+        tuple_type: ArbitraryTuplType,
+        depth: u32,
+    },
+    /// The tuple type was not recognised by the parser
+    TupleTypeUnrecognised,
+}
+
+impl Display for DecoderError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            DecoderError::PnmMagicInvalid(magic) =>
+                f.write_fmt(format_args!("Expected magic constant for PNM: P1..P7, got [{:#04X?}, {:#04X?}]", magic[0], magic[1])),
+            DecoderError::UnparsableValue(src, data, err) =>
+                f.write_fmt(format_args!("Error parsing {:?} as {}: {}", data, src, err)),
+
+            DecoderError::NonAsciiByteInHeader(c) =>
+                f.write_fmt(format_args!("Non-ASCII character {:#04X?} in header", c)),
+            DecoderError::NonAsciiLineInPamHeader =>
+                f.write_str("Non-ASCII line in PAM header"),
+            DecoderError::NonAsciiSample =>
+                f.write_str("Non-ASCII character where sample value was expected"),
+
+            DecoderError::NotNewlineAfterP7Magic(c) =>
+                f.write_fmt(format_args!("Expected newline after P7 magic, got {:#04X?}", c)),
+            DecoderError::UnexpectedPnmHeaderEnd =>
+                f.write_str("Unexpected end of PNM header"),
+
+            DecoderError::HeaderLineDuplicated(line) =>
+                f.write_fmt(format_args!("Duplicate {} line", line)),
+            DecoderError::HeaderLineUnknown(identifier) =>
+                f.write_fmt(format_args!("Unknown header line with identifier {:?}", identifier)),
+            DecoderError::HeaderLineMissing { height, width, depth, maxval } =>
+                f.write_fmt(format_args!("Missing header line: have height={:?}, width={:?}, depth={:?}, maxval={:?}", height, width, depth, maxval)),
+
+            DecoderError::InputTooShort =>
+                f.write_str("Not enough data was provided to the Decoder to decode the image"),
+            DecoderError::UnexpectedByteInRaster(c) =>
+                f.write_fmt(format_args!("Unexpected character {:#04X?} within sample raster", c)),
+            DecoderError::SampleOutOfBounds(val) =>
+                f.write_fmt(format_args!("Sample value {} outside of bounds", val)),
+            DecoderError::MaxvalTooBig(maxval) =>
+                f.write_fmt(format_args!("Image MAXVAL exceeds {}: {}", 0xFFFF, maxval)),
+
+            DecoderError::InvalidDepthOrMaxval { tuple_type, depth, maxval } =>
+                f.write_fmt(format_args!("Invalid depth ({}) or maxval ({}) for tuple type {}", depth, maxval, tuple_type.name())),
+            DecoderError::InvalidDepth { tuple_type, depth } =>
+                f.write_fmt(format_args!("Invalid depth ({}) for tuple type {}", depth, tuple_type.name())),
+            DecoderError::TupleTypeUnrecognised =>
+                f.write_str("Tuple type not recognized"),
+        }
+    }
+}
+
+impl From<DecoderError> for ImageError {
```

Comment:
Could you add a note here, saying that this needs to be hidden if `pnm` becomes a separate crate, until it is `1.0`? It works for private types though, which is very nice.

Reply:
Done